### PR TITLE
A0-2083: `cliain` pipeline for Liminal

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
# Description

1. We build `cliain` image (for standard node) for every PR. We **do not** push the image though.
1. We build `cliain` image (for liminal node) for every PR. We **do not** push the image though. We also run tests for SNARK relations.
1. We build and push `cliain` image (for standard node) for every push and workflow dispatch.
1. We build and push `cliain` image (for liminal node) for every push and workflow dispatch.

Tags and Docker registry:
 - for standard node, everything is kept as before
 - for liminal node, we are using [cliain-liminal](https://eu-central-1.console.aws.amazon.com/ecr/repositories/public/573243519133/cliain-liminal) with commit sha as tag (we intentionally don't want `latest`)

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: .github/scripts/test_cliain_snark_relations.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+${CLIAIN} snark-relation generate-proof -p deposit.groth16.pk.bytes deposit \
+  --note "2257517311045912551,9329547706917600007,17678219388335595033,2758194574870438734" \
+  --token-id 1 \
+  --token-amount 10 \
+  --trapdoor 17 \
+  --nullifier 19
+${CLIAIN} snark-relation verify \
+  --verifying-key-file deposit.groth16.vk.bytes \
+  --proof-file deposit.groth16.proof.bytes \
+  --public-input-file deposit.groth16.public_input.bytes
```
Comment (deuszx) [inline .github/scripts/test_cliain_snark_relations.sh:34]: And what exactly are we testing here? i.e. what would have to happen for this script to exit with an error and fail the CI job?

Comment (deuszx) [reply .github/scripts/test_cliain_snark_relations.sh:34]: Ah, I see that you added `exit(1)` to the verify command in the `cliain`.

File: .github/workflows/build-and-push-cliain.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            ${{ matrix.runtime == 'standard' &&
+                  (steps.get_branch.outputs.branch_name == 'main' && 'latest' ||
+                  steps.get_branch.outputs.branch_name) ||
+                matrix.runtime == 'liminal' && steps.get_branch.outputs.sha_short }}
+        uses: docker/build-push-action@v3.3.0
+        with:
+          context: ./bin/cliain
+          builder: ${{ steps.buildx.outputs.name }}
+          file: ./bin/cliain/Dockerfile
+          push: ${{ github.event_name == 'workflow_dispatch' || github.event_name == 'push' }}
```
Comment (deuszx) [inline .github/workflows/build-and-push-cliain.yml:75]: That's gonna build and publish on every push ... to `main`? 

Also, what is _workflow dispatch_?

Comment (pmikolajczyk41) [reply .github/workflows/build-and-push-cliain.yml:75]: 1. yes, so image pushing is done for, and only for pushing to `main`, building will be additionally done for every commit in PR
2. manual trigger
image

**Review by mikogs — APPROVED** [review]

**Review by deuszx — APPROVED** [review]

**Pull request merged**